src/backend/distributed/metadata/metadata_cache.c

```diff
@@ -2193,8 +2198,37 @@ CitusCatalogNamespaceId(void)
 Oid
 DistObjectRelationId(void)
 {
-	CachedRelationNamespaceLookup("pg_dist_object", CitusCatalogNamespaceId(),
-								  &MetadataCache.distObjectRelationId);
+	/*
+	 * In older versions pg_dist_object was living in the `citus` namespace, With Citus 11
+	 * this has been moved to pg_dist_catalog.
+	 *
+	 * During downgrades it could therefore be that we simply need to look in the new
+	 * catalog. Since we expect to find it most of the time in the citus schema from
+	 * now on we will start there.
+	 *
+	 * even after the table has been moved, the oid's stay the same, so we don't have to
+	 * invalidate the cache after a move
+	 *
+	 * Note: during testing we also up/downgrade the extension, and sometimes interact
+	 * with the database when the schema and the binary are not in sync. Hance we always
+	 * allow the catalog to be missing on our first lookup. The error message might
+	 * therefore become misleading as it will complain about citus.pg_dist_object not
+	 * being found when called too early.
+	 */
+	CachedRelationNamespaceLookupExtended("pg_dist_object",
+										  CitusCatalogNamespaceId(),
+										  &MetadataCache.distObjectRelationId,
+										  true);
+	if (!OidIsValid(MetadataCache.distObjectRelationId))
+	{
+		/*
+		 * We can only ever reach here while we are creating/altering our extension before
+		 * the table is moved back to citus.
+		 */
```

Comment:
Wouldn't our version checks guard against this problem? Or are we adding this specifically for our test suite that runs at certain moments with the checks disabled.

The comments are quite clear on what we are doing, but I am lacking the context of the why. Would be good to capture here when this condition holds true as I was under the impression that version checks would guard against doing anything before altering the extension to the right version for the binary.